### PR TITLE
E2E: Handle SSH failures during machine log collection 

### LINKED ISSUE #997: Add Azure boot diagnostics as an e2e log collection strategy
/kind feature

**Describe the solution you'd like**

While the implementation of the CAPI e2e test framework `ClusterLogCollector` in #976 is working, it relies on SSH connectivity. Sometimes that [fails](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/988/pull-cluster-api-provider-azure-e2e/1316355750150279168), which is unfortunate because those cases are more likely to have logs of interest.

Perhaps the e2e log collector could have a fallback strategy that does the equivalent of `az vm boot-diagnostics get-boot-log`, or maybe even use the serial console to collect the expected set of logs.

Failing that, it might be worth adding retry code around the `Dial` calls in the test helper func `execOnHost`.

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Handles SSH failures during log collection by:
- Retrying remote execution of commands
- Collecting boot logs from azure boot diagnostics

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #997 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Handle SSH failures during machine log collection 
```

### CODE REVIEW

**Comment by mboersma** [comment]
/lgtm

Nice work! I see that we got `boot.log` for each instance in this CI run as well. 👍🏼

File: test/e2e/retry.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0.1
+	retryBackoffSteps           = 3
+)
+
+// retryWithExponentialBackOff retries the function until it returns nil,
+// or until the number of attempts (steps) has reached the maximum value.
+func retryWithExponentialBackOff(fn func() error) error {
```
Comment (nader-ziada) [inline test/e2e/retry.go:36]: nice! @shysank

**Comment by nader-ziada** [comment]
/lgtm
/approve

/hold 
@CecileRobertMichon did you want to wait merging PRs for the patch release?

File: test/e2e/retry.go (hunk 1 above)
Comment (shysank) [reply test/e2e/retry.go:36]: Thanks! Although I was heavily inspired (shamelessly copied) by [this](https://github.com/kubernetes-sigs/cluster-api/blob/e6b0039b4aa1021b0a9afec8a527d969c20acabb/cmd/clusterctl/client/cluster/client.go#L238) :)

Comment (nader-ziada) [reply test/e2e/retry.go:36]: lol, that's okay, you still used it nicely

**Comment by CecileRobertMichon** [comment]
nope, I cherry-pick the PRs I need 

/hold cancel

**Pull request merged**